pkg/kubectl/cluster.go

```diff
@@ -30,6 +31,8 @@ import (
 type ClusterGeneratorV1Beta1 struct {
 	// Name of the cluster context (required)
 	Name string
+	// Name of the federation this cluster is joining (required)
```

Comment:
> On a side note, do we have a grid for which versions of kubefed, the FCP, the host cluster Kubernetes version and the joined cluster versions we need to support, in new federation and upgrade cases? I know that we want to maintain backwards compatibility, but I'm not sure I really understand what sort of compatibility we're trying to maintain.

We have neither concretely defined nor documented that right now. I guess that will be one of the important work items in taking federation to GA.

But having said that, I think @nikhiljindal's comment holds. I did not think about that before.

> Adding the annotation does give you the ability to get all objects related to a particular federation, rather than having to get all the annotated objects and then the cluster objects separately.

You need to do that anyway. You need to get all the annotated objects from the host cluster and then separately get the cluster objects from the federation API server irrespective of the annotations. Even if your objects are annotated, I believe you still need to get each type of object separately and filter for objects that contain the annotation you care about. Our API machinery does not allow getting objects with a given annotation across different API types.

@nikhiljindal, having said all that, I don't think adding the annotation here is that dangerous. This required param is an internal detail and @perotinus is making it mandatory to establish the contract between the provider and the client. The client here is the kubefed join implementation. Users who run the `kubefed join` command don't have to care about this. So it is not going to break them. It is just that the implementation of kubefed join should provide these annotation values which it does.